Plutarch/Rational.hs

```diff
@@ -44,120 +47,142 @@ import Plutarch.Builtin (
   pfromRational :: Term s (PRational :--> a)
 
 data PRational s
-  = PRational (Term s PInteger) (Term s PInteger)
+  = PRational (Term s PInteger) (Term s PPositive)
   deriving stock (Generic)
-  deriving anyclass (PlutusType)
+  deriving anyclass (PlutusType, PEq)
+
 instance DerivePlutusType PRational where type DPTStrat _ = PlutusTypeScott
 
 instance (PNum a, PFractional a) => Fractional (Term s a) where
   (/) = (#/)
   recip x = precip # x
-  fromRational x = pfromRational #$ pcon $ PRational (pconstant (numerator x)) (pconstant (denominator x))
+  fromRational x =
+    pfromRational #$ pcon $
+      PRational
+        (pconstant $ numerator x)
+        (punsafeDowncast . pconstant $ denominator x)
 
 instance PShow PRational where
   pshow' _ x =
     pshowRat # x
     where
       pshowRat = phoistAcyclic $
         plam $ \n -> pmatch n $ \(PRational x y) ->
-          pshow x <> "/" <> pshow y
+          pshow x <> "/" <> pshow (pto y)
 
 instance PIsData PRational where
-  pfromDataImpl x' = phoistAcyclic (plam $ \x -> pListToRat #$ pmap # pasInt #$ pasList # pforgetData x) # x'
+  pfromDataImpl x' = phoistAcyclic (plam $ \x -> plistToRat #$ pasList # pforgetData x) # x'
+    where
+      plistToRat :: Term s (PBuiltinList PData :--> PRational)
+      plistToRat = plam $ \x ->
+        pcon $
+          PRational (pasInt #$ phead # x)
+            . punsafeDowncast
+            $ pasInt #$ phead #$ ptail # x
   pdataImpl x' =
     phoistAcyclic
-      ( plam $ \x ->
-          punsafeCoerce (pdata $ pRatToList # x :: Term _ (PAsData (PBuiltinList (PAsData PInteger))))
+      ( plam $ \x -> unTermCont $ do
+          PRational a b <- tcont $ pmatch x
+          let res :: Term _ (PBuiltinList (PAsData PInteger))
+              res = pcons # pdata a #$ pcons # pdata (pto b) #$ pnil
+          pure $ pdataImpl res
       )
       # x'
 
-pRatToList :: Term s (PRational :--> PBuiltinList (PAsData PInteger))
-pRatToList = plam $ \x -> pmatch x $ \(PRational a b) ->
-  pcons # pdata a #$ pcons # pdata b #$ punsafeCoerce (pnil :: Term s (PBuiltinList PData))
+newtype Flip f a b = Flip (f b a) deriving stock (Generic)
 
-pListToRat :: Term s (PBuiltinList PInteger :--> PRational)
-pListToRat = plam $ \x -> pcon $ PRational (phead # x) (phead #$ ptail # x)
-
-instance PEq PRational where
-  l' #== r' =
-    phoistAcyclic
-      ( plam $ \l r ->
-          pmatch l $ \(PRational ln ld) ->
-            pmatch r $ \(PRational rn rd) ->
-              rd * ln #== rn * ld
-      )
-      # l'
-      # r'
+-- | NOTE: This instance produces a verified 'PPositive' as the excess output.
+instance PTryFrom PData (PAsData PRational) where
+  type PTryFromExcess PData (PAsData PRational) = Flip Term PPositive
+  ptryFrom' opq = runTermCont $ do
+    (_, ld) <- tcont $ ptryFrom @(PAsData (PBuiltinList PData)) opq
+    tcont $ \f -> pif (plength # ld #== 2) (f ()) (ptraceError "ptryFrom(PRational): data list length should be 2")
```

Comment:
This isn't efficient, but we can leave it for later.

Reply:
Add a TODO

Reply:
I optimized it